internal/alb/rs/rules.go

```diff
@@ -3,202 +3,339 @@ package rs
-}
+	max := len(current)
+	if len(desired) > max {
+		max = len(desired)
+	}
+
+	for i := 1; i <= max; i++ {
+		is := fmt.Sprintf("%v", i)
+		c := currentMap[is]
+		d := desiredMap[is]
 
-// FindByPriority returns the position in the Rules slice of the rule parameter
-func (r Rules) FindByPriority(priority *string) (int, *Rule) {
-	for p, v := range r {
-		if v.rs.current == nil {
-			continue
+		if c == nil && d != nil {
+			add = append(add, d)
 		}
-		if awsutil.DeepEqual(v.rs.current.Priority, priority) {
-			return p, v
+		if c != nil && d != nil {
+			d.RuleArn = c.RuleArn
+			modify = append(modify, d)
+		}
+		if c != nil && d == nil {
+			remove = append(remove, c)
 		}
 	}
-	return -1, nil
-}
 
-// FindUnusedTGs returns a list of TargetGroups that are no longer referncd by any of
-// the rules passed into this method.
-func (r Rules) FindUnusedTGs(tgs tg.TargetGroups, defaultArn *string) tg.TargetGroups {
-	var unused tg.TargetGroups
+	return
+}
 
-TG:
-	for _, t := range tgs {
-		used := false
+func (c *rulesController) getDesiredRules(ingress *extensions.Ingress, targetGroups tg.TargetGroups) ([]*Rule, error) {
+	var output []*Rule
 
-		arn := t.CurrentARN()
-		if arn == nil {
-			continue
+	currentPriority := 1
+	for _, rule := range ingress.Spec.Rules {
+		if len(rule.HTTP.Paths) == 0 {
+			return nil, fmt.Errorf("Ingress doesn't have any paths defined. This is not a very good Ingress")
 		}
 
-		if defaultArn != nil && *arn == *defaultArn {
-			continue
+		for _, path := range rule.HTTP.Paths {
+			r := &Rule{
+				Rule: elbv2.Rule{
+					Priority: aws.String(fmt.Sprintf("%v", currentPriority)),
+				},
+				Backend: path.Backend,
+			}
+
+			// Handle the annotation based actions
+			if action.Use(path.Backend.ServicePort.String()) {
+				annos, err := c.store.GetIngressAnnotations(k8s.MetaNamespaceKey(ingress))
+				if err != nil {
+					return nil, err
+				}
+				actionConfig, err := annos.Action.GetAction(path.Backend.ServiceName)
+				if err != nil {
+					return nil, err
+				}
+				r.Actions[0] = actionConfig
+			} else {
+				i := targetGroups.LookupByBackend(path.Backend)
+				if i < 0 {
+					return nil, fmt.Errorf("unable to locate a target group for backend %v:%v",
+						path.Backend.ServiceName, path.Backend.ServicePort.String())
+				}
+
+				r.Actions = []*elbv2.Action{{
+					Type:           aws.String(elbv2.ActionTypeEnumForward),
+					TargetGroupArn: targetGroups[i].CurrentARN(),
+				}}
+			}
+
+			if rule.Host != "" {
+				r.Conditions = append(r.Conditions, condition("host-header", rule.Host))
+			}
+
+			if path.Path != "" {
+				r.Conditions = append(r.Conditions, condition("path-pattern", path.Path))
+			}
+
+			output = append(output, r)
+			currentPriority++
 		}
+	}
+	return output, nil
+}
 
-		for _, rule := range r {
-			if rule.rs.current == nil {
-				continue TG
+func (c *rulesController) getCurrentRules(listenerArn string) (results []*Rule, err error) {
+	p := request.Pagination{
```

Comment:
I'd rather not deal with terrible aws pagination api in application logic(which makes the unit test hard).
Instead, i'd prefer to create a wrapper api in elbv2 that returns a list of rules directly.